relay-server/src/extractors/decoder.rs

```diff
@@ -195,22 +215,26 @@ impl Stream for DecodingPayload {
     type Error = <SharedPayload as Stream>::Error;
 
     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
-        let encoded = match self.payload.poll() {
-            Ok(Async::Ready(Some(encoded))) => encoded,
-            Ok(Async::Ready(None)) => return Ok(Async::Ready(None)),
-            Ok(Async::NotReady) => return Ok(Async::NotReady),
-            Err(error) => return Err(error),
-        };
+        loop {
```

Comment:
This is the largest change. To summarize the diff:
- If the payload returns `None`, finish the decoder. This may return `None` again or another chunk of bytes; both is fine at that point
- If the decoder doesn't yield a result, continue looping to poll the next chunk